contracts/PrizeDistributionFactoryV2.sol

```diff
@@ -0,0 +1,272 @@
+    {
+        return _calculatePrizeDistribution(_drawId);
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+     * @notice Calculate Prize Distribution for a given drawId
+     * @param _drawId Draw ID
+     * @return PrizeDistribution
+     */
+    function _calculatePrizeDistribution(uint32 _drawId)
+        internal
+        view
+        virtual
+        returns (IPrizeDistributionBuffer.PrizeDistribution memory)
+    {
+        IPrizeTierHistoryV2.PrizeTierV2 memory _prizeTier = prizeTierHistory.getPrizeTier(_drawId);
+        IDrawBeacon.Draw memory _draw = drawBuffer.getDraw(_drawId);
+
+        (
+            uint64[] memory _startTimes,
+            uint64[] memory _endTimes
+        ) = _calculateDrawPeriodTimestampOffsets(
+                _draw.timestamp,
+                _draw.beaconPeriodSeconds,
+                _prizeTier.endTimestampOffset
+            );
+
+        uint256 _totalSupply = ticket.getAverageTotalSuppliesBetween(_startTimes, _endTimes)[0];
+
+        (uint8 _cardinality, uint104 _numberOfPicks) = _calculateCardinalityAndNumberOfPicks(
+            _prizeTier.bitRangeSize,
+            _prizeTier.prize,
+            _prizeTier.dpr,
+            minPickCost,
+            _totalSupply
+        );
+
+        IPrizeDistributionBuffer.PrizeDistribution
+            memory _prizeDistribution = IPrizeDistributionSource.PrizeDistribution({
+                bitRangeSize: _prizeTier.bitRangeSize,
+                matchCardinality: _cardinality,
+                startTimestampOffset: _draw.beaconPeriodSeconds,
+                endTimestampOffset: _prizeTier.endTimestampOffset,
+                maxPicksPerUser: _prizeTier.maxPicksPerUser,
+                expiryDuration: _prizeTier.expiryDuration,
+                numberOfPicks: _numberOfPicks,
+                tiers: _prizeTier.tiers,
+                prize: _prizeTier.prize
+            });
+
+        return _prizeDistribution;
+    }
+
+    /**
+     * @notice Compute prize pool cardinality and number of picks for a draw.
+     * @dev `cardinality` must be gte to one, that's why we use a do/while loop to increase it.
+     * @param _bitRangeSize Bit range size
+     * @param _prize Total prize amount
+     * @param _dpr Draw percentage rate
+     * @param _minPickCost Minimum cost for a pick
+     * @param _totalSupply Prize Pool Ticket total supply
+     * @return cardinality and number of picks
+     */
+    function _calculateCardinalityAndNumberOfPicks(
+        uint8 _bitRangeSize,
+        uint256 _prize,
+        uint32 _dpr,
+        uint256 _minPickCost,
+        uint256 _totalSupply
+    ) internal pure returns (uint8 cardinality, uint104 numberOfPicks) {
+        uint256 _odds = (_dpr * _totalSupply) / _prize;
+
+        if (_odds == 0) {
+            return (cardinality = 1, numberOfPicks);
+        }
+
+        uint256 _targetPicks = ((_totalSupply / _minPickCost) * RATE_NORMALIZATION) / _odds;
```

Comment:
There is significant loss here because you are dividing before you multiply

remember: try to avoid any kind of loss of precision.  When you divide you lose precision.  By multiplying after the loss of precision **you are magnifying the imprecision**. 

Instead, multiply *before* you divide. When you multiply **before** dividing, the loss is minimized.

For example, assume:
_dpr = 0.00005479452054794521 * 1e9 = 54794
_totalSupply = 1000
_prize = 100

=> _odds = 547940

_totalSupply = 100
_minPickCost = 80
RATE_NORMALIZATION = 1e9

Then:

((_totalSupply / _minPickCost) * RATE_NORMALIZATION) / _odds = ((1000 / 80) * 1e9) / 547940
= (12 * 1e9) / 547940
= 21900

vs

(_totalSupply * RATE_NORMALIZATION) / (_odds * _minPickCost) = (100 * 1e9 / (547940 * 80)
= 22812

Reply:
You could go one step further as well:

uint256 _targetPicks = (_totalSupply * RATE_NORMALIZATION * _prize) / (_minPickCost * _dpr * _totalSupply)

by combining the odds calc.

Reply:
Refactored in the following commit: https://github.com/pooltogether/v4-periphery/pull/54/commits/6282f4c374743553722fd0cf85760ce3d64030bd